File: mobile/android/app/src/main/java/com/quickdrop/app/webview/JSBridge.kt
package com.quickdrop.app.webview

import android.content.Intent
import android.net.Uri
import android.webkit.JavascriptInterface
import android.webkit.WebView
import com.quickdrop.app.QuickDropApplication
import com.quickdrop.app.camera.QRScannerActivity
import org.json.JSONArray

/**
 * JavaScript ↔ Android Native 桥接
 *
 * 暴露给 WebView 中 JS 调用的原生能力：
 * - 扫码
 * - Token 存取
 * - 文件选择（通过系统文件选择器，支持多选）
 * - 前台服务控制
 *
 * Web 端调用方式：window.QuickDropBridge.<method>()
 */
class JSBridge(private val webView: WebView) {

    private val tokenManager = QuickDropApplication.instance.tokenManager

    /**
     * 文件选择回调 — 由 MainActivity 注入
     * 用于触发系统文件选择器（需要 Activity 级别的 ActivityResultLauncher）
     */
    var filePickerCallback: (() -> Unit)? = null

    /**
     * 打开扫码页面
     * JS 调用: QuickDropBridge.startQRScanner()
     */
    @JavascriptInterface
    fun startQRScanner() {
        val intent = Intent(webView.context, QRScannerActivity::class.java)
        webView.context.startActivity(intent)
    }

    /**
     * 保存 Access Token
     * JS 调用: QuickDropBridge.saveToken(accessToken, refreshToken)
     */
    @JavascriptInterface
    fun saveToken(accessToken: String, refreshToken: String) {
        tokenManager.saveTokens(accessToken, refreshToken)
    }

    /**
     * 获取 Access Token（异步返回给 JS）
     * JS 调用: QuickDropBridge.getAccessToken()
     * 返回 token 字符串，无则返回空
     */
    @JavascriptInterface
    fun getAccessToken(): String {
        return tokenManager.getAccessToken() ?: ""
    }

    /**
     * 获取 Refresh Token
     */
    @JavascriptInterface
    fun getRefreshToken(): String {
        return tokenManager.getRefreshToken() ?: ""
    }

    /**
     * 清除所有 Token
     * JS 调用: QuickDropBridge.clearTokens()
     */
    @JavascriptInterface
    fun clearTokens() {
        tokenManager.clearTokens()
    }

    /**
     * 获取认证服务地址
     */
    @JavascriptInterface
    fun getAuthBaseUrl(): String {
        return com.quickdrop.app.BuildConfig.AUTH_BASE_URL
    }

    /**
     * 获取信令服务 WebSocket 地址
     */
    @JavascriptInterface
    fun getSignalWsUrl(): String {
        return com.quickdrop.app.BuildConfig.SIGNAL_WS_URL
    }

    /**
     * 保存接收到的文件到本地存储
     * JS 调用: QuickDropBridge.saveReceivedFile(fileName, base64Data)
     *
     * 将 base64 数据解码写入缓存目录，然后弹出系统分享面板让用户选择保存位置
     */
    @JavascriptInterface
    fun saveReceivedFile(fileName: String, base64Data: String) {
        try {
            val bytes = android.util.Base64.decode(base64Data, android.util.Base64.DEFAULT)
            val cacheDir = webView.context.cacheDir
            val file = java.io.File(cacheDir, fileName)
            file.writeBytes(bytes)

            // 通过 FileProvider 获取 URI，启动分享 Intent
            val uri = androidx.core.content.FileProvider.getUriForFile(
                webView.context,
                "${webView.context.packageName}.fileprovider",
                file
            )
            val shareIntent = android.content.Intent(android.content.Intent.ACTION_SEND).apply {
                type = "*/*"
                putExtra(android.content.Intent.EXTRA_STREAM, uri)
                addFlags(android.content.Intent.FLAG_GRANT_READ_URI_PERMISSION)
                addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            webView.context.startActivity(android.content.Intent.createChooser(shareIntent, "保存文件"))
        } catch (e: Exception) {
            android.util.Log.e("JSBridge", "saveReceivedFile failed", e)
        }
    }

    /**
     * 选择文件（触发 Android 系统文件选择器，支持多选）
     * JS 调用: QuickDropBridge.pickFiles()
     *
     * 依赖 MainActivity 注入的 filePickerCallback 来启动 ActivityResultLauncher。
     * 结果通过 onFilesPicked() 回调给 JS 的 onNativeFilePicked(filePathsJson)。
     */
    @JavascriptInterface
    fun pickFiles() {
        filePickerCallback?.invoke()
    }

    /**
     * 文件选择完成回调（由 MainActivity 在收到文件选择结果后调用）
     * @param filePaths 可访问的文件路径列表（已从 content URI 复制到临时目录）
     */
    fun onFilesPicked(filePaths: List<String>) {
        val jsonArray = JSONArray(filePaths)
        val escaped = escapeForJS(jsonArray.toString())
        webView.post {
            webView.evaluateJavascript(
                "if(typeof onNativeFilePicked === 'function') onNativeFilePicked('$escaped')",
                null
            )
        }
    }

    /**
     * JS 字符串转义 — 用于安全地注入 JS 代码
     * 对应 iOS JSBridge.escapeForJS()
     */
    private fun escapeForJS(str: String): String {
        return str
            .replace("\\", "\\\\")
            .replace("'", "\\'")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
    }

    /**
     * 页面加载完成后注入已保存的 Token 到 JS
     */
    fun injectToken() {
        val accessToken = tokenManager.getAccessToken()
        val refreshToken = tokenManager.getRefreshToken()
        if (accessToken != null && refreshToken != null) {
            webView.evaluateJavascript(
                """
                if (typeof onNativeTokensReady === 'function') {
                    onNativeTokensReady('$accessToken', '$refreshToken');
                }
                """.trimIndent(),
                null
            )
        }
    }

    /**
     * 从 JS 接收扫码结果
     * 由 QRScannerActivity 在扫码成功后调用
     */
    fun onQRCodeScanned(qrData: String) {
        webView.post {
            val escaped = qrData.replace("'", "\\'").replace("\n", "\\n")
            webView.evaluateJavascript(
                "if(typeof onQRCodeScanned === 'function') onQRCodeScanned('$escaped')",
                null
            )
        }
    }
}

File: mobile/ios/QuickDrop/Bridge/JSBridge.swift
import Foundation
import WebKit

/// JavaScript ↔ iOS Native 桥接
///
/// 通过 WKScriptMessageHandler 接收来自 WebView 中 JS 的调用。
/// Web 端调用方式：window.webkit.messageHandlers.quickdrop.postMessage({action: "...", ...})
///
/// 暴露给 WebView 的原生能力：
/// - Token 存取 (Keychain)
/// - 扫码 (AVFoundation)
/// - 文件选择 (UIDocumentPickerViewController)
/// - 服务地址获取
///
/// 对应 Android 的 JSBridge (JavascriptInterface)
final class JSBridge: NSObject, WKScriptMessageHandler {

    private weak var webView: WKWebView?
    private weak var viewController: UIViewController?
    private let tokenManager = TokenManager.shared

    init(webView: WKWebView, viewController: UIViewController) {
        self.webView = webView
        self.viewController = viewController
        super.init()
    }

    // MARK: - WKScriptMessageHandler

    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard message.name == "quickdrop",
              let body = message.body as? [String: Any],
              let action = body["action"] as? String else {
            return
        }

        switch action {
        case "saveTokens":
            handleSaveTokens(body)
        case "clearTokens":
            handleClearTokens()
        case "startQRScanner":
            handleStartQRScanner()
        case "pickFiles":
            handlePickFiles()
        case "getAuthBaseUrl":
            handleGetAuthBaseUrl()
        case "getSignalWsUrl":
            handleGetSignalWsUrl()
        case "saveReceivedFile":
            handleSaveReceivedFile(body)
        default:
            print("[JSBridge] Unknown action: \(action)")
        }
    }

    // MARK: - Action Handlers

    private func handleSaveTokens(_ body: [String: Any]) {
        let accessToken = body["accessToken"] as? String ?? ""
        let refreshToken = body["refreshToken"] as? String ?? ""
        tokenManager.saveTokens(accessToken: accessToken, refreshToken: refreshToken)
    }

    private func handleClearTokens() {
        tokenManager.clearTokens()
    }

    private func handleStartQRScanner() {
        guard let vc = viewController else { return }
        let scanner = QRScannerViewController()
        scanner.onScanResult = { [weak self] qrData in
            self?.onQRCodeScanned(qrData: qrData)
        }
        scanner.modalPresentationStyle = .fullScreen
        vc.present(scanner, animated: true)
    }

    private func handlePickFiles() {
        guard let vc = viewController else { return }
        let picker = UIDocumentPickerViewController(forOpeningContentTypes: [.data, .movie, .image, .audio, .pdf, .text])
        picker.allowsMultipleSelection = true
        picker.delegate = self
        vc.present(picker, animated: true)
    }

    private func handleSaveReceivedFile(_ body: [String: Any]) {
        guard let fileName = body["fileName"] as? String,
              let base64 = body["data"] as? String,
              let fileData = Data(base64Encoded: base64) else {
            print("[JSBridge] saveReceivedFile: invalid data")
            return
        }

        // 保存到临时目录
        let tempDir = FileManager.default.temporaryDirectory
        let fileURL = tempDir.appendingPathComponent(fileName)
        do {
            try fileData.write(to: fileURL)
            print("[JSBridge] File saved to: \(fileURL.path)")
        } catch {
            print("[JSBridge] Failed to save file: \(error)")
            return
        }

        // 在主线程弹出分享面板，让用户选择保存位置
        DispatchQueue.main.async { [weak self] in
            guard let vc = self?.viewController else { return }
            let activityVC = UIActivityViewController(activityItems: [fileURL], applicationActivities: nil)
            // iPad 适配
            if let popover = activityVC.popoverPresentationController {
                popover.sourceView = vc.view
                popover.sourceRect = CGRect(x: vc.view.bounds.midX, y: vc.view.bounds.midY, width: 0, height: 0)
            }
            vc.present(activityVC, animated: true)
        }
    }

    private func handleGetAuthBaseUrl() {
        injectJS("window.__QUICKDROP_CONFIG__ = { ...window.__QUICKDROP_CONFIG__, authBaseUrl: '\(AppConfig.authBaseUrl)' }")
    }

    private func handleGetSignalWsUrl() {
        injectJS("window.__QUICKDROP_CONFIG__ = { ...window.__QUICKDROP_CONFIG__, signalWsUrl: '\(AppConfig.signalWsUrl)' }")
    }

    // MARK: - JS Injection (Native → Web)

    /// 页面加载完成后注入已保存的 Token 到 JS
    func injectTokens() {
        guard let accessToken = tokenManager.getAccessTokenRaw(),
              let refreshToken = tokenManager.getRefreshToken() else {
            return
        }
        let escapedAccess = escapeForJS(accessToken)
        let escapedRefresh = escapeForJS(refreshToken)
        injectJS("""
        if (typeof onNativeTokensReady === 'function') {
            onNativeTokensReady('\(escapedAccess)', '\(escapedRefresh)');
        }
        """)
    }

    /// 注入配置信息到 JS
    func injectConfig() {
        let authUrl = escapeForJS(AppConfig.authBaseUrl)
        let wsUrl = escapeForJS(AppConfig.signalWsUrl)
        injectJS("""
        window.__QUICKDROP_CONFIG__ = {
            authBaseUrl: '\(authUrl)',
            signalWsUrl: '\(wsUrl)'
        };
        """)
    }

    /// 扫码结果回调给 JS
    func onQRCodeScanned(qrData: String) {
        let escaped = escapeForJS(qrData)
        injectJS("if(typeof onQRCodeScanned === 'function') onQRCodeScanned('\(escaped)')")
    }

    /// 文件选择结果回调给 JS
    func onFilesPicked(fileUris: [String]) {
        guard let jsonData = try? JSONSerialization.data(withJSONObject: fileUris),
              let jsonString = String(data: jsonData, encoding: .utf8) else { return }
        let escaped = escapeForJS(jsonString)
        injectJS("if(typeof onNativeFilePicked === 'function') onNativeFilePicked('\(escaped)')")
    }

    // MARK: - Private Helpers

    private func injectJS(_ js: String) {
        DispatchQueue.main.async { [weak self] in
            self?.webView?.evaluateJavaScript(js)
        }
    }

    /// 转义字符串用于 JS 单引号字符串内
    private func escapeForJS(_ string: String) -> String {
        return string
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "'", with: "\\'")
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: "\r", with: "\\r")
    }
}

// MARK: - UIDocumentPickerDelegate

extension JSBridge: UIDocumentPickerDelegate {
    func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        // 将选中的文件复制到 App 临时目录，确保 WebRTC 可读取
        var accessibleUrls: [String] = []
        for url in urls {
            guard url.startAccessingSecurityScopedResource() else { continue }
            defer { url.stopAccessingSecurityScopedResource() }

            // 复制到临时目录
            let tempDir = FileManager.default.temporaryDirectory
            let destUrl = tempDir.appendingPathComponent(url.lastPathComponent)
            try? FileManager.default.copyItem(at: url, to: destUrl)
            accessibleUrls.append(destUrl.path)
        }
        onFilesPicked(fileUris: accessibleUrls)
    }
}
